web3swift/src/Contract/Statically Typed/ABIDecoder+Static.swift

```diff
@@ -10,81 +10,137 @@ import Foundation
-            return EthereumAddress(data)
-        }
-        
-        guard let bytes = data.bytesFromHex else { throw ABIError.invalidValue }
-        
-        guard let decodedData = try ABIDecoder.decode(bytes, forType: ABIRawType.FixedAddress, offset: 0) as? String else {
-            throw ABIError.invalidValue
-        }
-        
-        return EthereumAddress(decodedData)
+    public static func decode(_ data: ParsedABIEntry, to: EthereumAddress.Type) throws -> EthereumAddress {
+        return EthereumAddress(data)
     }
     
-    public static func decode(_ data: String, to: BigInt.Type) throws -> BigInt {
+    public static func decode(_ data: ParsedABIEntry, to: BigInt.Type) throws -> BigInt {
         guard let value = BigInt(hex: data) else { throw ABIError.invalidValue }
         return value
     }
     
-    public static func decode(_ data: String, to: BigUInt.Type) throws -> BigUInt {
+    public static func decode(_ data: ParsedABIEntry, to: BigUInt.Type) throws -> BigUInt {
         guard let value = BigUInt(hex: data) else { throw ABIError.invalidValue }
         return value
     }
     
-    public static func decode(_ data: String, to: UInt8.Type) throws -> UInt8 {
+    public static func decode(_ data: ParsedABIEntry, to: UInt8.Type) throws -> UInt8 {
         guard let value = BigUInt(hex: data) else { throw ABIError.invalidValue }
         guard value.bitWidth <= 8 else { throw ABIError.invalidValue }
         return UInt8(value)
     }
     
-    public static func decode(_ data: String, to: UInt16.Type) throws -> UInt16 {
+    public static func decode(_ data: ParsedABIEntry, to: UInt16.Type) throws -> UInt16 {
         guard let value = BigUInt(hex: data) else { throw ABIError.invalidValue }
         guard value.bitWidth <= 16 else { throw ABIError.invalidValue }
         return UInt16(value)
     }
     
-    public static func decode(_ data: String, to: UInt32.Type) throws -> UInt32 {
+    public static func decode(_ data: ParsedABIEntry, to: UInt32.Type) throws -> UInt32 {
         guard let value = BigUInt(hex: data) else { throw ABIError.invalidValue }
         guard value.bitWidth <= 32 else { throw ABIError.invalidValue }
         return UInt32(value)
     }
     
-    public static func decode(_ data: String, to: UInt64.Type) throws -> UInt64 {
+    public static func decode(_ data: ParsedABIEntry, to: UInt64.Type) throws -> UInt64 {
         guard let value = BigUInt(hex: data) else { throw ABIError.invalidValue }
         guard value.bitWidth <= 64 else { throw ABIError.invalidValue }
         return UInt64(value)
     }
     
-    public static func decode(_ data: String, to: URL.Type) throws -> URL {
-        // If from log value, already decoded during initial log decode process
-        let filtered = data.stringValue.trimmingCharacters(in: CharacterSet(charactersIn: "\0"))
+    public static func decode(_ data: [ParsedABIEntry], to: [EthereumAddress].Type) throws -> [EthereumAddress] {
```

Comment:
Too specific -> generic?

Reply:
Hm requires a bit more thought and work as I can't find a way to manage with container of types, maybe abstracting further. I'll try to fix it.